File: src/frontend/src/lib/api-models/trusted-partner.ts
import type {
  ListTrustedPartnersResponse as ApiListTrustedPartnersResponse,
  CreateTrustedPartnerRequest as ApiCreateTrustedPartnerRequest,
  CreateTrustedPartnerResponse as ApiCreateTrustedPartnerResponse,
  TrustedPartner as ApiTrustedPartner,
} from '@ssn/backend-api';

export type ListTrustedPartnersResponse = TrustedPartner[];

export type CreateTrustedPartnerRequest = {
  principal: string;
  name: string;
};

export type CreateTrustedPartnerResponse = TrustedPartner;

export type TrustedPartner = {
  id: string;
  principal: string;
  name: string;
};

export function mapListTrustedPartnersResponse(
  res: ApiListTrustedPartnersResponse,
): ListTrustedPartnersResponse {
  return res.map(mapTrustedPartnerResponse);
}

export function mapCreateTrustedPartnerRequest(
  req: CreateTrustedPartnerRequest,
): ApiCreateTrustedPartnerRequest {
  return {
    principal_id: req.principal,
    name: req.name,
  };
}

export function mapCreateTrustedPartnerResponse(
  res: ApiCreateTrustedPartnerResponse,
): CreateTrustedPartnerResponse {
  return mapTrustedPartnerResponse(res);
}

export function mapTrustedPartnerResponse(
  res: ApiTrustedPartner,
): TrustedPartner {
  return {
    id: res.id,
    principal: res.principal_id,
    name: res.name,
  };
}

File: src/backend/src/mapping/trusted_partner.rs
use crate::{
    data::{self},
    dto::{CreateTrustedPartnerRequest, ListTrustedPartnersResponse, TrustedPartner},
};
use candid::Principal;

pub fn map_list_trusted_partners_response(
    trusted_partners: Vec<(data::Uuid, data::TrustedPartner)>,
) -> ListTrustedPartnersResponse {
    trusted_partners
        .into_iter()
        .map(|(id, trusted_partner)| map_trusted_partner_response(id, trusted_partner))
        .collect()
}

pub fn map_trusted_partner_response(
    id: data::Uuid,
    trusted_partner: data::TrustedPartner,
) -> TrustedPartner {
    TrustedPartner {
        id: id.to_string(),
        name: trusted_partner.name,
        principal_id: trusted_partner.principal.to_string(),
    }
}

pub fn map_create_trusted_partner_request(
    req: CreateTrustedPartnerRequest,
) -> Result<data::TrustedPartner, String> {
    Ok(data::TrustedPartner {
        name: req.name,
        principal: Principal::from_text(&req.principal_id)
            .map_err(|err| format!("Failed to convert principal {}: {}", req.principal_id, err))?,
    })
}
